src/shared/util/logic.ts

```diff
@@ -201,7 +202,7 @@ export const getLogicUnitPreventingSubmit = (
   const preventSubmitConditions = getPreventSubmitConditions(form)
   return preventSubmitConditions.find((logicUnit) =>
     isLogicUnitSatisfied(
-      submission,
+      formatFieldsForLogic(submission, form.form_fields),
```

Comment:
it seems we are taking on a lot of additional complexity because the adapter layer is in the logic module itself, and as a result the adapter functions have to account for both client and server shapes. might it be better to shift the adapter layers outside instead, and have a separate adapter for both client and server so that they can evolve independently?